### PR TITLE
Override Flag Defaults with Environment Variables

### PR DESCRIPTION
**Pull request by kwd42**
**Purpose**: I am creating systemd scripts to start reading a meter at bootup on my C.H.I.P. I think the best practice is to have the configuration set through environment variables rather than via command line arguments. Thus, the service would read the configuration from /etc/default/rtlamr. The second part of the change is to add a new flag to allow for appending to an existing log file rather than overwrite it.

**Solution**: Before processing the command line arguments, read in the expected environment variables. This would override the hard coded default values. These would optionally be overridden by values entered on the command line. The expected environment variable names are simply the option name set to upper case with "RTLAMR_" prefix. The second part of the change simply adds a `-logappend=<bool>` option.

**Discussion**: I am not opposed to changing EV names if the prefix is not desired, for example.

_Note_: This is my first use of git and github, so please let me know if I have not followed proper protocol.

### CODE REVIEW

File: flags.go [hunk 1]
```diff
@@ -106,11 +110,27 @@ func RegisterFlags() {
 	}
 }
 
+func EnvVarFlags() {
+	var value string
+
+	for f := range rtlamrFlags {
+		value = os.Getenv("RTLAMR_" + strings.ToUpper(f));
+		if 0<len(value) {
```
Comment (bemasher) [inline flags.go]: This should probably be `if value != "" {`. The docs state that looking up environment variables which are not set will return an empty string.

File: flags.go [hunk 2]
```diff
@@ -106,11 +110,27 @@ func RegisterFlags() {
...
+			flag.Set(f, value)
```
Comment (bemasher) [inline flags.go]: We need to handle errors produced here when values stored in environment variables aren't valid.

File: help.md [hunk 3]
```diff
@@ -2,6 +2,7 @@
 Detailed usage information for the various flags of RTLAMR.
 
   - `logfile` writes log statements to the given file. Defaults to `/dev/stdout`.
+  - `logappend` append to, instead of replacing, logfile at startup (only valid when logfile is not /dev/stdout)
```
Comment (bemasher) [inline help.md]: Please revert this change, this file will likely be removed soon because it is out of date.

File: flags.go [hunk 4]
```diff
@@ -48,6 +48,8 @@ var timeLimit = flag.Duration("duration", 0, "time to run for, 0 for infinite, e
 var meterID MeterIDFilter
 var meterType MeterTypeFilter
 
+var logAppend = flag.Bool("logappend", false, "apppend to log file if it exists")
```
Comment (bemasher) [inline flags.go]: Please remove this flag, new log output handling flags are not something I want to deal with. At some point I'm going to remove all logging output related flags since console redirection should be sufficient.

File: docs/index.md [hunk 5]
```diff
@@ -39,6 +39,7 @@ Usage of rtlamr:
   -format=plain: format to write log messages in: plain, csv, json, xml or gob
   -gobunsafe=false: allow gob output to stdout
   -logfile=/dev/stdout: log statement dump file
+  -logappend=false: append to, instead of replacing, logfile at startup
```
Comment (bemasher) [inline docs/index.md]: Please revert this change, `docs/index.md` is kept in sync with `README.md` manually and I will update it at the next release.

File: README.md [hunk 6]
```diff
@@ -31,6 +31,7 @@ Usage of rtlamr:
   -format=plain: format to write log messages in: plain, csv, json, xml or gob
   -gobunsafe=false: allow gob output to stdout
   -logfile=/dev/stdout: log statement dump file
+  -logappend=false: append to, instead of replacing, logfile at startup
```
Comment (bemasher) [inline README.md]: Please remove `-logappend`, see later comments.

File: flags.go [hunk 7]
```diff
@@ -106,11 +110,27 @@ func RegisterFlags() {
 	}
 }
 
+func EnvVarFlags() {
+	var value string
+
+	for f := range rtlamrFlags {
```
Comment (bemasher) [inline flags.go]: Instead of looping over `rtlamrFlags`, it may be better to be more general and loop over all flags registered prior to calling this function, this will catch rtl_tcp flags as well. `flag.VisitAll` is a more appropriate way to check for environment variables matching a flag rtlamr knows about.

https://godoc.org/flag#FlagSet.VisitAll

File: flags.go [hunk 8]
```diff
@@ -59,32 +61,34 @@ var single = flag.Bool("single", false, "one shot execution, if used with -filte
...
+	"filtertype":   true,
+	"format":       true,
+	"gobunsafe":    true,
+	"quiet":        true,
+	"unique":       true,
+	"single":       true,
+	"cpuprofile":   true,
+	"fastmag":      true,
+	"version":      true,
+}
```
Comment (bemasher) [inline flags.go]: This should be moved back to `RegisterFlags()`, see later comments.

**Comment by bemasher** [comment]
I've implemented environment variable flag overriding functionality separately. I am closing this pull request because it includes a commit which I have since removed that causes some significant performance issues while manipulating the repository with git. You've been credited with the feature suggestion in 9e258fdd9e231e2df02ccb3394572e7bf6454554.

**Comment by kwd42** [comment]
OK. Thanks. I will change my systemd script to make use of redirection.

I was originally going to, as a separate branch, add the various etc files used for the systemd scripts, such as /etc/default/rtlamr with in a new etc folder on the project. Would you rather I not do that and make a separate project instead?

**Comment by bemasher** [comment]
It may be best to keep that as a separate project. I've considered having a contrib folder as part of the git repo but I'd rather not have to be gate-keeper for updates to it.
